Add a post-deploy hook sample that cleans up kubernetes resources that were deployed by Cloud Deploy

menahyouyeah opened the pull request:
Add a post-deploy hook sample that cleans up kubernetes resources that were deployed by Cloud Deploy. 

*Includes a QUICKSTART and sample config files to use this post-deploy hook.

mattsanta reviewed (commented):
1) WDYT of the structure used by custom targets for each sample? i.e.
- quickstart/
- image implementation/ 
- README

2) remove "sample" from the directory name, it's redundant since this entire repo is samples.

3) Looking at the implementation really makes me think we need to start pushing towards introducing packages that can be shared across our sample image implementation.  Just from this PR...
- This likely won't be the last time we have a sample that uses kubectl or gcloud. Can we introduce packages that expose wrappers for them? Maybe take a look at how we use kubectl in the cd image.
- Same with respect to metadata, we will likely have more samples that would need to reimplement what's here unless we create a package for it.
- Every sample is going to need to hardcode the env var keys Cloud Deploy sets in the container unless we have a shared package that exposes them all.

I feel like it might be more work to move towards separate packages later rather than now given how, for example, the kubectl implementation is very specific to this sample. If we introduce a package for a kubectl client then it needs to be pretty generic so it can be used by other samples.

Would it be easier to first add all the shared packages and then the cleanup sample?

mattsanta commented:
Oh also forgot, we need to set up the tests to run. I think Angel did a bit of this with GitHub actions for vertex AI.

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/QUICKSTART.md:
```diff
@@ -0,0 +1,144 @@
+* deployment.apps
+* replicaset.apps
+* statefulset.apps
+* pod,configmap
```
typo my bad

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/.dockerignore:
```diff
@@ -0,0 +1 @@
+*_test.go
```
Removed

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/QUICKSTART.md:
```diff
@@ -0,0 +1,144 @@
+# Kubernetes Resource Clean Up
+This contains a sample container that can be used to clean up Kubernetes
+resources that were deployed by Cloud Deploy. It should be used as a post-deploy
```
Done

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/QUICKSTART.md:
```diff
@@ -0,0 +1,144 @@
+At a high level, the sample image:
+1. Gets a list of kubernetes resources that were deployed by Cloud Deploy in the 
+   **current** release
+2. Gets a list of **all** kubernetes resources that were deployed by Cloud Deploy
```
added a line on the filtering done

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/QUICKSTART.md:
```diff
@@ -0,0 +1,144 @@
+
+There are two command line flags you can pass to the container:
+
+1. `namespace`: Namespace(s) to filter on when finding resources to delete. For 
```
Done

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/kubectl_test.go:
```diff
@@ -0,0 +1,86 @@
+	os.Setenv(pipelineEnvKey, "mypipeline")
+	os.Setenv(targetEnvKey, "mytarget")
+	os.Setenv(projectEnvKey, "myproject")
+	os.Setenv(locationEnvKey, "losangeles")
```
Done

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/main.go:
```diff
@@ -0,0 +1,99 @@
+var gkeClusterRegex = regexp.MustCompile("^projects/([^/]+)/locations/([^/]+)/clusters/([^/]+)$")
+
+const (
+	// The name of the post-deploy hook cleanup sample, this is passed back to
```
why though? Do we want to keep that confidential?

mattsanta replied:
Not necessarily, but I don't think we need to advertise that it's used for metrics tracking.

menahyouyeah replied:
Ok done

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/kubectl.go:
```diff
@@ -0,0 +1,187 @@
+)
+
+var (
+	defaultResources = []string{
```
Removed

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/kubectl.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+// listResources returns a list of resources, filtered by resource type, namespace,
+// and were deployed by Cloud Deploy. If includeReleaseLabel is true, it filtered
```
Done

menahyouyeah commented on post-deploy-hooks/k8s-cleanup-sample/QUICKSTART.md:
```diff
@@ -0,0 +1,144 @@
+REGION, and IMAGE values or update your existing Cloud Deploy config
+file to reference a post-deploy hook, and your Skaffold file to then reference
+the image you built.
+3. If using the samples, create a GKE cluster (note this can take 10 min to run):
```
I updated the QUICKSTART to do 2 releases so they can actually see the post-deploy hook in action, and updated the README.

menahyouyeah commented:
> 1. WDYT of the structure used by custom targets for each sample? i.e.
> 
> * quickstart/
> * image implementation/
> * README

I did all of the above except putting the actual implementation code in a separate folder. It just didn't seem necessary yet to nest? I can change if you feel strongly though. 

I addressed all other comments, and updated the github action trigger to run tests for the clean-up sample. Per our discussion offline, I will begin the refactor after this PR, in smaller separate PRs.
